library/kotlin/src/io/envoyproxy/envoymobile/GRPCResponseHandler.kt

```diff
@@ -0,0 +1,177 @@
+   *
+   * @param closure: Closure which will be called when an error occurs.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onError(closure: (error: EnvoyError) -> Unit): GRPCResponseHandler {
+    this.errorClosure = closure
+    underlyingHandler.onError(closure)
+    return this
+  }
+
+  /**
+   * Recursively processes a buffer of data, buffering it into messages based on state.
+   * When a message has been fully buffered, `onMessage` will be called with the message.
+   *
+   * @param bufferedStream The buffer of data from which to determine state and messages.
+   * @param processState The current process state of the buffering.
+   * @param onMessage Closure to call when a new message is available.
+   */
+  private fun processData(
+      bufferedStream: ByteArrayOutputStream,
+      processState: ProcessState,
+      onMessage: (byteBuffer: ByteBuffer) -> Unit): ProcessState {
+
+    var nextState = processState
+
+    when (processState) {
+      is ProcessState.CompressionFlag -> {
+        val byteArray = bufferedStream.toByteArray()
+        if (byteArray.isEmpty()) {
+          // We don't have enough information to extract the compression flag, so we'll just return
+          return ProcessState.CompressionFlag
+        }
+
+        val compressionFlag = byteArray[0]
+        // TODO: Support gRPC compression https://github.com/lyft/envoy-mobile/issues/501
+        if (compressionFlag.compareTo(0) != 0) {
+          errorClosure(
+              EnvoyError(
+                  EnvoyErrorCode.ENVOY_UNDEFINED_ERROR,
+                  "Unable to accept compression enabled messages for gRPC"))
+
+          // no op the current onData and clean up
+          underlyingHandler.onData { _, _ -> }
+          bufferedStream.reset()
+        }
+
+        nextState = ProcessState.MessageLength
+      }
+      is ProcessState.MessageLength -> {
+        if (bufferedStream.size() < GRPC_PREFIX_LENGTH) {
+          // We don't have enough information to extract the message length, so we'll just return
+          return ProcessState.MessageLength
+        }
+
+        val byteArray = bufferedStream.toByteArray()
+        val messageLength = ByteBuffer.wrap(byteArray.sliceArray(1..4)).int
+
+        if (messageLength == 0) {
+          onMessage(ByteBuffer.wrap(ByteArray(0)))
+          return ProcessState.CompressionFlag
+        }
+
+        nextState = ProcessState.Message(messageLength)
+      }
+      is ProcessState.Message -> {
+        if (bufferedStream.size() < processState.messageLength + GRPC_PREFIX_LENGTH) {
+          // We don't have enough bytes to construct the message, so we'll just return
+          return ProcessState.Message(processState.messageLength)
+        }
+
+        val byteArray = bufferedStream.toByteArray()
+        onMessage(ByteBuffer.wrap(
+            byteArray.sliceArray(GRPC_PREFIX_LENGTH until GRPC_PREFIX_LENGTH + processState.messageLength)))
```

Comment:
Curious, why use `byteArray.sliceArray(1..4)` above but `sliceArray(... until ...)` here?

Reply:
Just checked, we aren't allowed to use variables with the `..` operator, instead we are supposed to use `until`